### DISCUSSION TITLE
Problem with definition of conditional return types

### DISCUSSION THREAD

**Discussion by Itatsi** [vimeo/psalm#9890] [Q&A] [answered]
Hello people

I have not been using psalm for very long and I was hoping you could give me a tip on how to properly handle the error Psalm identified.

Thanks a lot in advance^^

https://psalm.dev/r/1b651c8b27
```
<?php

class a {
    /** @var ?array{id:int,...} */
	private ?array $data = null;
    
    public function load(int $a):bool {  
        $this->data = ['id' => $a,'b'=> ''];
        return true;
    }
    public function fnc():int|false {
    	if (!isset($this->data['id'])) return false;
        
        return 1;
    }
}

function test(int $a):int{return $a;}

$foo = new a();
if ($foo->load(1)) {
    test($foo->fnc());
}
```

```
Psalm output (using commit [5df772b](https://github.com/vimeo/psalm/commit/5df772b)): 

ERROR: [PossiblyFalseArgument](https://psalm.dev/104) - 22:10 - Argument 1 of test cannot be false, possibly int value expected
```

**Comment by orklah** [comment, marked as answer]
In theory, this is what you want:
https://psalm.dev/r/191c5d72a9

I made two changes: 
- I added an assertion on load() that affirms that fnc() will return an int if load() returned true
- I documented the class as  immutable

The last part is required by Psalm because it's not safe to try to assert types on methods that are subject to be changed.

However, your class is very much not immutable because you save things in a property.

That makes the whole pattern pretty risky in terms of static analysis (for example, you could imagine calling load a second time with an invalid input that would completely change the first assertion or you could call another method on $foo that could have an effect on $this->data and change the assumption)

So Psalm is not letting you do it. There's no safe way to document your code as is
